### PR TITLE
master(dm): check subtask status when `stop-task` instead of exit quickly

### PR DESCRIPTION
**Pull request by fatelei**
### What problem does this PR solve? 

#3254 

### What is changed and how it works?

I check the code: when master stop task, the task will remove subtask from subtask holder.
so when master check subtask status if the subtask status is not wanted, the check process continue.

### Check List 

Tests 

 - fix unit tests

### Release note 

```release-note
NONE
```

### CODE REVIEW

**Comment by lance6716** [comment]
thanks for your contribution, can you add a test like https://github.com/pingcap/ticdc/blob/77029891ff124a5a0e07d5b4dbd4c4f40c6c6268/dm/dm/master/server_test.go#L518 to make sure this behaviour won't be changed in future?

I think the test can check that, dm-worker returns a subtask status with error for the first time, `waitOperationOk` will retry, dm-worker returns no subtask status for the second time, `waitOperationOk` will return successfully.

**Comment by fatelei** [comment]
> thanks for your contribution, can you add a test like
> 
> https://github.com/pingcap/ticdc/blob/77029891ff124a5a0e07d5b4dbd4c4f40c6c6268/dm/dm/master/server_test.go#L518
> 
> to make sure this behaviour won't be changed in future?
> I think the test can check that, dm-worker returns a subtask status with error for the first time, `waitOperationOk` will retry, dm-worker returns no subtask status for the second time, `waitOperationOk` will return successfully.

ok

**Comment by fatelei** [comment]
i am not sure why jenkins-ticdc verify job failed

**Comment by lance6716** [comment]
@fatelei 

```
FAIL: server_test.go:576: testMaster.TestStopTaskWithExceptRight
server_test.go:611:
    ok, msg, _, err := s.waitOperationOk(ctx, mockWorker, "", "", req)
/nfs/cache/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:183:
    h.t.Fatalf(format, args...)
... Error: Unexpected call to *pbmock.MockWorkerClient.QueryStatus([context.Background.WithDeadline(2021-12-05 16:20:04.024583066 +0800 CST m=+78.233373355 [944.968281ms]) ]) at /home/jenkins/agent/workspace/atom-ut/ticdc/dm/dm/master/workerrpc/rawgrpc.go:110 because: 
expected call at /home/jenkins/agent/workspace/atom-ut/ticdc/dm/dm/master/server_test.go:605 has already been called the max number of times
```

**Comment by fatelei** [comment]
> @fatelei
> 
> ```
> FAIL: server_test.go:576: testMaster.TestStopTaskWithExceptRight
> server_test.go:611:
>     ok, msg, _, err := s.waitOperationOk(ctx, mockWorker, "", "", req)
> /nfs/cache/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:183:
>     h.t.Fatalf(format, args...)
> ... Error: Unexpected call to *pbmock.MockWorkerClient.QueryStatus([context.Background.WithDeadline(2021-12-05 16:20:04.024583066 +0800 CST m=+78.233373355 [944.968281ms]) ]) at /home/jenkins/agent/workspace/atom-ut/ticdc/dm/dm/master/workerrpc/rawgrpc.go:110 because: 
> expected call at /home/jenkins/agent/workspace/atom-ut/ticdc/dm/dm/master/server_test.go:605 has already been called the max number of times
> ```

where can i see the error stack? click the link, can not see the details

File: dm/dm/master/server_test.go [hunk 1]
```diff
@@ -573,6 +573,47 @@ func (t *testMaster) TestWaitOperationOkRightResult(c *check.C) {
...
+	mockWorkerClient := pbmock.NewMockWorkerClient(ctrl)
+	mockWorkerClient.EXPECT().QueryStatus(
+		gomock.Any(),
+		gomock.Any(),
+	).Return(notExceptResp, nil).Return(exceptResp, nil).MaxTimes(2)
+	mockWorker := scheduler.NewMockWorker(newMockRPCClient(mockWorkerClient))
+
+	ok, msg, _, err := s.waitOperationOk(ctx, mockWorker, taskName, "", req)
+	c.Assert(err, check.IsNil)
+	c.Assert(ok, check.Equals, "ok")
```
Comment (lance6716) [inline dm/dm/master/server_test.go:613]: you can use (..., check.IsTrue) and (..., check.HasLen, 0)

File: dm/dm/master/server_test.go [hunk 2]
```diff
@@ -573,6 +573,47 @@ func (t *testMaster) TestWaitOperationOkRightResult(c *check.C) {
...
+				Status: &pb.SubTaskStatus_Sync{Sync: &pb.SyncStatus{
+					UnresolvedGroups: []*pb.ShardingGroup{{Target: "`db`.`tbl`", Unsynced: []string{"table1"}}},
+				}},
+			},
+		},
+	}
+	exceptResp := &pb.QueryStatusResponse{SubTaskStatus: []*pb.SubTaskStatus{
+		{
+			Name:   "task1",
+			Status: &pb.SubTaskStatus_Msg{Msg: common2.NoSubTaskMsg(taskName)},
```
Comment (Ehco1996) [inline dm/dm/master/server_test.go:591]: i am confused with this test, why we have so many task names?  `test-stop-task` ,`task1`, `task-unittest`

Comment (lance6716) [reply dm/dm/master/server_test.go:591]: they should be one name to imitate the real case.

File: dm/dm/master/server_test.go [hunk 3]
```diff
@@ -573,6 +573,47 @@ func (t *testMaster) TestWaitOperationOkRightResult(c *check.C) {
...
+	}}
+
+	req := &pb.OperateTaskRequest{
+		Op:   pb.TaskOp_Stop,
+		Name: "task-unittest",
+	}
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	ctx := context.Background()
+	duration, _ := time.ParseDuration("1s")
```
Comment (lance6716) [inline dm/dm/master/server_test.go:602]: you can use `time.Second`

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: dm/dm/master/server_test.go [hunk 4]
```diff
@@ -573,6 +573,41 @@ func (t *testMaster) TestWaitOperationOkRightResult(c *check.C) {
...
+		SubTaskStatus: []*pb.SubTaskStatus{
+			{
+				Name: "task1",
+				Status: &pb.SubTaskStatus_Sync{Sync: &pb.SyncStatus{
+					UnresolvedGroups: []*pb.ShardingGroup{{Target: "`db`.`tbl`", Unsynced: []string{"table1"}}},
+				}},
+			},
+		},
+	}
+	exceptResp := &pb.QueryStatusResponse{SubTaskStatus: []*pb.SubTaskStatus{}}
```
Comment (lance6716) [inline dm/dm/master/server_test.go:588]: https://github.com/pingcap/ticdc/blob/9da1946522cb7d132ebc16138ef84746aff2e237/dm/dm/worker/status.go#L72-L76

why remove this response by git force push?

Comment (fatelei) [reply dm/dm/master/server_test.go]: misunderstand, unittest refactored

**Review by lance6716 — APPROVED** [review]

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**